cmd/google_cloud_ops_agent_diagnostics/main.go

```diff
@@ -22,17 +22,17 @@ import (
 )
 
 func getUnifiedConfigAndValidate(userConfPath, platform string) (confgenerator.UnifiedConfig, error) {
-	_, mergedConfig, err := confgenerator.MergeConfFiles(userConfPath, platform, apps.BuiltInConfStructs)
+	uc, err := confgenerator.MergeConfFiles(userConfPath, platform, apps.BuiltInConfStructs)
 	if err != nil {
 		return confgenerator.UnifiedConfig{}, err
 	}
 
-	uc, err := confgenerator.ParseUnifiedConfigAndValidate(mergedConfig, platform)
+	err = uc.Validate(platform)
```

Comment:
So I actually think the behaviour now is strictly better than before. I didn't see any good reason why the empty list for `include_paths` was ever prohibited, it doesn't break anything. Our validation on the field says it is `required` to be valid, not that it must be non-empty. If we do want that as well, we should add that as a validator and not rely on marshalling with `omitempty` and then unmarshalling again to validate this one validation rule.

Also the behaviour before was kinda broken. It would surface errors on the merged config and give yaml line numbers and show the error on the merged config. Which is never visible to the users anyway, so the errors were less than ideal too. The merged config should be an internal thing, and users should only interact with the user specified config.

Let me know if you think I should add a validator to ensure the old behaviour regarding us requiring non-empty `include_paths` (which we should also document) or whether we should loosed the requirements as we do here.

Reply:
YAML and validation are orthogonal. That we inline validation with unmarshalling is a convenience and not a requirement: it is absolutely possible to run struct validation entirely on its own without having to marshal/unmarshal anything to/from YAML.

The marshal/unmarshal round-trip should be a no-op. YAML is only an intermediate representation and should be state-preserving. The YAML tags for `include_paths` allow valid things to become invalid after a round-trip test, therefore `omitempty` was the wrong tag to use. If the original intent was that `include_paths` must not be empty then that should have been encoded entirely within the `validate` tag, e.g. `min=1` instead of `required`.

What is the use case for `include_paths` being empty? "That it doesn't break things" isn't the bar we should meet; the bar should at least be "we don't want the user to do the wrong thing", and an empty `include_paths` is clearly the wrong thing IMO, and should be disallowed.

Regarding where to put the struct-level validation: I would go the other way and call `uc.Validate` from within `MergeConfFiles`, which has the additional benefit of simplifying all of the call sites. I can't think of a reason why the intermediate state after-merge-but-before-validate would be useful so it should be OK to combine those steps.

For the missing struct-level validation, I would still be OK with assuming that valid inputs always produce a valid output during `MergeConfFiles`, which implies that struct-level validation should be performed inside there and issue a `panic` on failure.

Reply:
Looks good 👍 We might also be able to get away with moving `original.Validate` to the panic as well, since `original` is just the built-in struct rather than something that's user-provided.